gesturelibrary/src/double.c

```diff
@@ -41,45 +58,199 @@ int set_on_double_tap(void (*listener)(const double_tap_t*)) {
     }
 }
 
-static void update_double_taps(double_tap_t* double_tap, const stroke_t* stroke, const tap_t* tap) {
-    double_tap->x = stroke->x;
-    double_tap->y = stroke->y;
-    double_tap->t = stroke->t;
+static void update_double_taps(const tap_t* tap, const touch_event_t* event) {
+    int null_index     = -1;  // last seen null or failed double_d spot
+    int complete_index = -1;  // last seen complete spot
+    int update         = 1;
+    int group_index    = -1;  // double tap with the closest time the completed tap, will group with this tap
+    // int group_number;
+
+    float x_diff;
+    float y_diff;
+    float time_diff;
+
+    // printf("printing tap_state: %d\n", tap->state);
+
+    if (tap->state == RECOGNIZER_STATE_COMPLETED) {
+        // printf("entered with tap x = %f\n", tap->x);
+        // if completed, compare it to all double_taps in double_d
+        // look for any possible ones and compare time and position
+
+        for (int i = 0; i < MAX_TOUCHES; i++) {
+            // printf("entered for loop with tap x = %f\n", tap->x);
+            double_tap_t d_tap = double_tap_d[i];
+
+            switch (d_tap.state) {
+            case RECOGNIZER_STATE_POSSIBLE:
+                // check position and time
 
-    switch (double_tap->state) {
-    case RECOGNIZER_STATE_IN_PROGRESS:
-        // look at the timestamp of the tap and the location
-        // if too far, move on to the next one
-        // if timestamp is too far, set this double to failed
+                // checking tap position to double tap position
 
-        if (tap->state == RECOGNIZER_STATE_COMPLETED) {
-            int dist = SQUARED_DIST(double_tap, tap);
+                x_diff    = (event->x - d_tap.x0) < 0 ? (event->x - d_tap.x0) * -1 : (event->x - d_tap.x0);
+                y_diff    = (event->y - d_tap.y0) < 0 ? (event->y - d_tap.y0) * -1 : (event->y - d_tap.y0);
+                time_diff = (event->t - d_tap.t0) < 0 ? (event->t - d_tap.t0) * -1 : (event->t - d_tap.t0);
 
-            if (double_tap->t - stroke->t > DOUBLE_DIFF) {  // if t_diff too big, this dTap has failed
-                double_tap->state = RECOGNIZER_STATE_FAILED;
-            } else if (dist < DOUBLE_DISTANCE) {
-                double_tap->state = RECOGNIZER_STATE_COMPLETED;
+                // printf("Printing possible case: x = %f\n", d_tap.x0);
+                // printf("Printing tap data: x = %f\n", tap->x);
+                // printf("Printing distances: x_diff = %f, y_diff = %f, time_diff = %f\n", x_diff, y_diff, time_diff);
+
+                if (x_diff < DOUBLE_MAX_X && y_diff < DOUBLE_MAX_Y) {
```

Comment:
I would prefer the acceptable positions to be a circle not a square.
Please us euclidean distance instead.